File: inflearn-algorithm-class/2023-2 알고리즘/Recursion/stringPermutation.py
import sys

def converseWeight(s):
    sum = 0
    for i in range(len(s)):
        sum += (ord(s[i]) - ord('a')) * (-1)**i
    
    if sum > 0:
        return 1
    else:
        return 0

def stringPermutation(s, start, end):
    if start == end:
        return converseWeight("".join(s))
    else:
        result = 0
        for i in range(start, end + 1):
            s[start], s[i] = s[i], s[start]
            result += stringPermutation(s, start+1, end)
        return result
    


case = int(sys.stdin.readline())
for _ in range(case):
    word = input()
    s = list(word)
    res = stringPermutation(s, 0 ,len(s) -1)
    print(res)

File: inflearn-algorithm-class/2023-2 알고리즘/Recursion/stringPermutation.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int converseWeight(string s) {
    int sum = 0;
    for (int i = 0; i < s.length(); i++) {
        if (i % 2 == 0) {
            sum = sum + (s[i] - 'a');
        }
        else {
            sum = sum - (s[i] - 'a');
        }
    }

    if (sum > 0) {
        return 1;
    } else {
        return 0;
    }
}

int stringPermutation(string s, int start, int end) {
    int range = end - start;
    if (range == 1) {
        return converseWeight(s);
    } else {
        int result = 0;
        for (int i = 0; i < range; i++) {
            swap(s[start], s[start+i]);
            result += stringPermutation(s, start + 1, end);
            swap(s[start], s[start+i]);
        }
        return result;
    }
}

int main() {
    int caseCount;
    cin >> caseCount;

    for (int i = 0; i < caseCount; i++) {
        string word;
        cin >> word;
        int res = stringPermutation(word, 0, word.length());
        cout << res << endl;
    }

    return 0;
}
